### PR TITLE
Add support for image classification

### LINKED ISSUE #173: Support for computer vision tasks
Thanks for the amazing library! It would be great to add support for computer vision tasks to 🤗 optimum!

In particular, it would be amazing to use the new inference pipelines to take advantage of ONNX for inference when using image classification models. 

This is partially selfishly motivated for use in (https://github.com/davanstrien/flyswot/) which is being used to 'deploy' transformer models via command line (running on CPU). I have previously had good success using onnx to speed things up but found the process of exporting onnx models, getting data ready for inference etc. not very fun. The API for pipelines looks like it will make this task much more enjoyable and flexible so it would be amazing to have support for image classification models in optimum.

### PR DESCRIPTION
**Pull request by regisss**
# What does this PR do?

Fixes #173. This PR enables:
- to use an image-classification pipeline
- to perform quantization/optimization with computer vision models

At the moment, static post-training quantization with percentile/entropy calibration may not work with some models due to an issue with ONNX Runtime.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/onnxruntime/quantization.py [hunk 1]
```diff
@@ -279,7 +279,7 @@ def export(
         quantization_config: QuantizationConfig,
         calibration_tensors_range: Optional[Dict[NodeName, Tuple[float, float]]] = None,
         use_external_data_format: bool = False,
-        quantization_preprocessor: Optional[QuantizationPreprocessor] = None,
+        preprocessor: Optional[QuantizationPreprocessor] = None,
```
Comment (philschmid) [inline optimum/onnxruntime/quantization.py:282]: wouldn't this break backward compatibility?

File: docs/source/onnxruntime/modeling_ort.mdx [hunk 2]
```diff
@@ -12,13 +12,13 @@ specific language governing permissions and limitations under the License.
 
 # Optimum Inference with ONNX Runtime
 
-Optimum is a utility package for building and running inference with accelerated runtime like ONNX Runtime. 
-Optimum can be used to load optimized models from the [Hugging Face Hub](hf.co/models) and create pipelines 
+Optimum is a utility package for building and running inference with accelerated runtime like ONNX Runtime.
+Optimum can be used to load optimized models from the [Hugging Face Hub](hf.co/models) and create pipelines
```
Comment (philschmid) [inline docs/source/onnxruntime/modeling_ort.mdx:16]: are you using some kind of lint/prettier? (Just asking)

File: docs/source/pipelines.mdx [hunk 3]
```diff
@@ -198,16 +199,16 @@ Below you can find two examples on how you could [`~onnxruntime.ORTOptimizer`] a
...
+ `ORTModelFor*` class.
 
 ```diff
 from transformers import AutoTokenizer, pipeline
 -from transformers import AutoModelForQuestionAnswering
 +from optimum.onnxruntime import ORTModelForQuestionAnswering
 
 -model = AutoModelForQuestionAnswering.from_pretrained("deepset/roberta-base-squad2")
-+model = ORTModelForQuestionAnswering.from_transformers("optimum/roberta-base-squad2")
++model = ORTModelForQuestionAnswering.from_pretrained("optimum/roberta-base-squad2")
```
Comment (philschmid) [inline docs/source/pipelines.mdx:211]: Thanks for fix!

File: tests/onnxruntime/test_modeling_ort.py [hunk 4]
```diff
@@ -286,15 +290,15 @@ def test_supported_transformers_architectures(self, *args, **kwargs):
 
     def test_load_vanilla_transformers_which_is_not_supported(self):
         with self.assertRaises(Exception) as context:
-            model = ORTModelForTokenClassification.from_pretrained("t5-small", from_transformers=Tru)
```
Comment (philschmid) [inline tests/onnxruntime/test_modeling_ort.py:289]: the idea was here to use a "architecture" which we don't support with `from_transformers`

File: tests/onnxruntime/test_modeling_ort.py [hunk 5]
```diff
@@ -357,15 +361,15 @@ def test_supported_transformers_architectures(self, *args, **kwargs):
 
     def test_load_vanilla_transformers_which_is_not_supported(self):
         with self.assertRaises(Exception) as context:
-            model = ORTModelForFeatureExtraction.from_pretrained("google/vit-base-patch16-224", from_transformers=Tru)
```
Comment (philschmid) [inline tests/onnxruntime/test_modeling_ort.py:360]: the idea was here to use a "architecture" which we don't support with `from_transformers`

**Review by philschmid — APPROVED** [review]
LGTM!🔥✅ I left a few comments. Mostly regarding breaking backward compatibility. 
Other than this awesome PR! 🔥

File: optimum/onnxruntime/modeling_ort.py [hunk 6]
```diff
@@ -730,3 +733,87 @@ def _prepare_attention_mask_for_generation(
         else:
             # Ensure attention mask is on the same device as the input IDs
             return torch.ones(inputs.shape[:2], dtype=torch.long, device=inputs.device)
+
+
+IMAGE_CLASSIFICATION_SAMPLE = r"""
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_ort.py:738]: A bit nitpicking, but isn't it an "example" more than a "sample"?

**Review by michaelbenayoun — APPROVED** [review]
Great, thanks a lot for adding this @regisss !
Left a few comments, but those are minor stuff, LGTM!

File: docs/source/onnxruntime/modeling_ort.mdx (hunk 2 above)
Comment (regisss) [reply docs/source/onnxruntime/modeling_ort.mdx:16]: I switched on the setting in VSCode to trim extra whitespaces. But I can revert this if you think it should not be done in this PR.

File: optimum/onnxruntime/modeling_ort.py (hunk 6 above)
Comment (regisss) [reply optimum/onnxruntime/modeling_ort.py:738]: I fully agree with you, I just copy-pasted this from other similar classes XD
I'm going to push a new commit to modify this :)

File: tests/onnxruntime/test_modeling_ort.py (hunk 4 above)
Comment (regisss) [reply tests/onnxruntime/test_modeling_ort.py:289]: Yep you're right, this is not the change I wanted to push actually XD
I wanted to add a "e" at `from_transformers=Tru`, I'm going to push this in a new commit.

File: optimum/onnxruntime/quantization.py (hunk 1 above)
Comment (regisss) [reply optimum/onnxruntime/quantization.py:282]: Yep it would, that's why I pushed a second commit to revert this. I think you did the review before I pushed this second commit, now this change should not appear anymore.

File: tests/onnxruntime/test_modeling_ort.py (hunk 5 above)
Comment (regisss) [reply tests/onnxruntime/test_modeling_ort.py:360]: Same here

File: docs/source/onnxruntime/modeling_ort.mdx (hunk 2 above)
Comment (JingyaHuang) [reply docs/source/onnxruntime/modeling_ort.mdx:16]: Hehe, I asked the same question before.

File: docs/source/pipelines.mdx [hunk 7]
```diff
@@ -12,7 +12,7 @@ specific language governing permissions and limitations under the License.
 
 # Optimum pipelines for inference
 
-The [`~pipelines.pipeline`] function makes it simple to use models from the [Model Hub](https://huggingface.co/models) for accelerated inference on a variety of tasks such as text classification.
+The [`~pipelines.pipeline`] function makes it simple to use models from the [Model Hub](https://huggingface.co/models) for accelerated inference on a variety of tasks such as question answering or image classification.
```
Comment (JingyaHuang) [inline docs/source/pipelines.mdx:15]:
```suggestion
The [`~pipelines.pipeline`] function makes it simple to use models from the [Model Hub](https://huggingface.co/models) for accelerated inference on a variety of tasks such as text classification, question answering, and image classification.
```
I think that it makes sense to list them all, as they are typical and quite different examples.

File: docs/source/pipelines.mdx [hunk 8]
```diff
@@ -198,16 +199,16 @@ Below you can find two examples on how you could [`~onnxruntime.ORTOptimizer`] a
 ## Transformers pipeline usage
 
 The [`~pipelines.pipeline`] function is just a light wrapper around the `transformers.pipeline` function to enable checks for supported tasks and additional features
-, like quantization and optimization. This being said you can use the `transformers.pipeline` and just replace your `AutoFor*` with the optimum 
- `ORTModelFor*` class. 
+, like quantization and optimization. This being said you can use the `transformers.pipeline` and just replace your `AutoFor*` with the optimum
```
Comment (JingyaHuang) [inline docs/source/pipelines.mdx:202]:
```suggestion
, like quantization and optimization. This being said you can use the `transformers.pipeline` and just replace your `AutoModelFor*` with the optimum
```

File: optimum/onnxruntime/optimization.py [hunk 9]
```diff
@@ -54,22 +55,23 @@ def from_pretrained(
         Returns:
             An instance of `ORTOptimizer`.
         """
-        tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)
+        preprocessor = get_preprocessor(model_name_or_path)
```
[the remark this answers is not in the document]
Comment (regisss) [reply optimum/onnxruntime/optimization.py:58]: Actually it is from what I see [here](https://github.com/huggingface/transformers/blob/b6a65ae52a6df67180be8f1e162a3d1829a51bc0/src/transformers/onnx/utils.py#L71).

Regarding a unified type called `Preprocessor`, I'm in favour of it but on the other hand it is not clear in Transformers either and we also have a different kind of preprocessor for quantization. So I feel like it will always be quite ambiguous.

[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/onnxruntime/optimization.py:58]: I agree with you on the second point, maybe explicit is better for the end-user. For the first point, I think we need to figure it out!

File: optimum/onnxruntime/optimization.py [hunk 10]
```diff
@@ -117,7 +119,7 @@ def export(
 
         # Export the model if it has not already been exported to ONNX IR
         if not onnx_model_path.exists():
-            export(self.tokenizer, self.model, self._onnx_config, self.opset, onnx_model_path)
+            export(self.preprocessor, self.model, self._onnx_config, self.opset, onnx_model_path)
```
Comment (JingyaHuang) [inline optimum/onnxruntime/optimization.py:122]: I think that it would be better to add the keyword arguments on, although the term `preprocessor` doesn't exist for older version of transformers. WDYT?

**Review by JingyaHuang — APPROVED** [review]
LGTM(G means great here). :>

Just added some small nits.

File: tests/onnxruntime/test_modeling_ort.py (hunk 4 above)
Comment (regisss) [reply tests/onnxruntime/test_modeling_ort.py:289]: Thinking about it a bit more, I should actually remove `from_transformers=True` here right @philschmid ? Or setting it to `False`

Comment (philschmid) [reply tests/onnxruntime/test_modeling_ort.py:289]: Yes!

File: optimum/onnxruntime/modeling_ort.py (hunk 6 above)
Comment (philschmid) [reply optimum/onnxruntime/modeling_ort.py]: The naming is coming from transformers

File: tests/onnxruntime/test_modeling_ort.py (hunk 4 above)
Comment (regisss) [reply tests/onnxruntime/test_modeling_ort.py:289]: Okay I understand better. There was a bug in all the `test_load_vanilla_transformers_which_is_not_supported`. Since it was `from_transformers=Tru` and not `from_transformers=True`, an exception was raised as expected. However, then, `self.assertTrue("Unrecognized configuration class", context.exception)` will always pass because it assesses the truthness of the 1st argument (`"Unrecognized configuration class"` here). I guess what we want to do here is to check if `"Unrecognized configuration class"` is in `context.exception`. I'm going to fix this.

Comment (regisss) [reply tests/onnxruntime/test_modeling_ort.py:289]: Also, for the feature extraction task, all models are supported (see [here](https://github.com/huggingface/transformers/blob/f47afefb210a120545a5e9745292aaf6c316d246/src/transformers/pipelines/feature_extraction.py#L15)) so I will just remove this test for this task.

File: docs/source/pipelines.mdx (hunk 7 above)
Comment (regisss) [reply docs/source/pipelines.mdx:15]: Actually all the supported tasks are listed a few lines below. My point was to show that it is not only NLP and not only classification (which is why I replaced "text classification" by "question answering"). But your suggestion is fine, I'm going to integrate it :)

File: optimum/onnxruntime/optimization.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (regisss) [reply optimum/onnxruntime/optimization.py:58]: Actually `AutoFeatureExtractor` and `AutoProcessor` will both return a `FeatureExtractorMixin` object.

File: optimum/onnxruntime/optimization.py (hunk 10 above)
Comment (michaelbenayoun) [reply optimum/onnxruntime/optimization.py:122]: The constraint would be transformers=4.7+ ?

Comment (regisss) [reply optimum/onnxruntime/optimization.py:122]: Keeping it as it is for the moment to not break backwards compatibility.

**Comment by davanstrien** [comment]
Thanks so much for doing this -- really happy to see this feature added 🤗

**Comment by regisss** [comment]
> Thanks so much for doing this -- really happy to see this feature added hugs

Do not hesitate to show us how you use it :) @davanstrien

**Comment by rhkenne** [comment]
Thank you very much for the great work. The description of this PR mentioned "perform quantization/optimization with computer vision models". Does this include CLIP model (CLIPProcessor, CLIPModel) ? I am looking for an approach to serialize CLIP models (to onnx) and perform quantization/optimization. Any idea how ?

**Comment by regisss** [comment]
> Thank you very much for the great work. The description of this PR mentioned "perform quantization/optimization with computer vision models". Does this include CLIP model (CLIPProcessor, CLIPModel) ? I am looking for an approach to serialize CLIP models (to onnx) and perform quantization/optimization. Any idea how ? @NielsRogge @JingyaHuang

@rhkenne The first step would be to add ONNX support for CLIP in Transformers. Would you be interested in opening a PR to enable this?

Then, for optimization and quantization, additional work would be necessary because multimodal models are not supported yet.

**Comment by regisss** [comment]
@rhkenne There is an ongoing PR [here](https://github.com/huggingface/transformers/pull/18515) to make CLIP exportable to ONNX.
